### PR TITLE
Add Support for Relaxed Data Binding

### PR DESCRIPTION
**Pull request by SachinAkash01**
Fixes: [#4870](https://github.com/ballerina-platform/ballerina-library/issues/4870)
See also: [#7366](https://github.com/ballerina-platform/ballerina-library/issues/7366)

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility
- [ ] Checked the impact on OpenAPI generation

### CODE REVIEW

File: ballerina/http_annotation.bal [hunk 1]
```diff
@@ -27,6 +27,7 @@
 # + validation - Enables the inbound payload validation functionality which provided by the constraint package. Enabled by default
 # + serviceType - The service object type which defines the service contract. This is auto-generated at compile-time
 # + basePath - Base path to be used with the service implementation. This is only allowed on service contract types
+# + laxDataBinding - Enables or disalbles relaxed data binding on the service side. Disabled by default
```
Comment (lnash94) [inline ballerina/http_annotation.bal:30]:
```suggestion
# + laxDataBinding - Enables or disables relaxed data binding on the service side. Disabled by default
```

File: ballerina/http_status_code_client.bal [hunk 2]
```diff
@@ -26,13 +26,15 @@ import ballerina/observe;
 #                HTTP service in resilient manner
 # + cookieStore - Stores the cookies of the client
 # + requireValidation - Enables the inbound payload validation functionalty which provided by the constraint package
+# + requireLaxDataBinding - Enables or disalbles relaxed data binding on the client side.
```
Comment (lnash94) [inline ballerina/http_status_code_client.bal:29]:
```suggestion
# + requireLaxDataBinding - Enables or disables relaxed data binding on the client side.
```

File: ballerina/http_client_payload_builder.bal [hunk 3]
```diff
@@ -159,18 +159,30 @@ isolated function jsonPayloadBuilder(Response response, TargetType targetType) r
...
 }
 
-isolated function nonNilablejsonPayloadBuilder(Response response, typedesc<anydata> targetType)
+isolated function nonNilablejsonPayloadBuilder(Response response, typedesc<anydata> targetType, boolean requireLaxDataBinding)
         returns anydata|ClientError {
     json payload = check response.getJsonPayload();
-    var result = jsondata:parseAsType(payload, {enableConstraintValidation: false, allowDataProjection: false}, targetType);
+    var result = jsondata:parseAsType(
+        payload,
+        requireLaxDataBinding
```
Comment (TharmiganK) [inline ballerina/http_client_payload_builder.bal:167]: Shall we move the options outside and define a variable for that:
```bal
jsondata:Options jsonParserOptions = {
    enableConstraintValidation: false,
    allowDataProjection: requireLaxDataBinding ? {nilAsOptionalField: true, absentAsNilableType: true} : false
};
```

I think this more readable

File: ballerina/http_types.bal [hunk 4]
```diff
@@ -105,6 +105,8 @@ public type CommonClientConfiguration record {|
     boolean validation = true;
     # Provides settings related to client socket configuration
     ClientSocketConfig socketConfig = {};
+    # Enables or disalbles relaxed data binding on the client side. Disabled by default
```
Comment (TharmiganK) [inline ballerina/http_types.bal:108]: Add more description in the documentation
```suggestion
    # Enables or disables relaxed data binding on the client side. Disabled by default.
    # When enabled, the JSON data will be projected to the Ballerina record type and
    # during the projection, nil values will be considered as optional fields and
    # absent fields will be considered for nilable types
```

File: ballerina/http_annotation.bal (hunk 1 above)
Comment (TharmiganK) [inline ballerina/http_annotation.bal:30]: Add more information about the data projection. Check the comment in the client config

File: native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java [hunk 5]
```diff
@@ -36,6 +36,9 @@
 import static io.ballerina.stdlib.http.api.HttpConstants.ALLOW_DATA_PROJECTION;
 import static io.ballerina.stdlib.http.api.HttpConstants.ENABLE_CONSTRAINT_VALIDATION;
 import static io.ballerina.stdlib.http.api.HttpConstants.PARSER_AS_TYPE_OPTIONS;
+import static io.ballerina.stdlib.http.api.HttpConstants.NIL_AS_OPTIONAL;
+import static io.ballerina.stdlib.http.api.HttpConstants.ABSENT_AS_NILABLE;
+
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/http/api/service/signature/converter/JsonToRecordConverter.java:41]: Lets remove this additional line

**Review by TharmiganK — APPROVED** [review]
LGTM

**Review by lnash94 — APPROVED** [review]

**Pull request merged**